### PR TITLE
simple feature escalated heavily, sorry for that

### PR DESCRIPTION
**Pull request by doemser**
- Has now a push to github feature
- has a useFetch hook based on zustand
- is now using mui with a custom theme
- has now a monaco code editor ( vscode )

Only thing I didnt touch is your proompting :D Better instruct gpt in english du reply in german, feels better

### CODE REVIEW

File: hooks/useFetch.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    isLoading: false,
+  },
+  updateWiki: {
+    data: null,
+    error: null,
+    isLoading: false,
+  },
+  fetch: async (route, params) => {
+    set((state) => ({
+      [route]: { ...state[route], isLoading: true, error: null },
```
Comment (Satttoshi) [inline hooks/useFetch.ts:29]: teh fuck is happening here with `...state[route]` ? :D

File: package.json [hunk 2]
```diff
@@ -11,15 +11,17 @@
   "dependencies": {
     "@emotion/react": "^11.11.1",
     "@emotion/styled": "^11.11.0",
+    "@monaco-editor/react": "^4.5.1",
```
Comment (Satttoshi) [inline package.json:14]: Nice VSC, on Website :D

File: pages/_app.tsx [hunk 3]
```diff
@@ -1,12 +1,38 @@
 import type { AppProps } from "next/app";
-import GlobalStyle from "../styles";
+import { ThemeProvider } from "@mui/material/styles";
+import CssBaseline from "@mui/material/CssBaseline";
+import { createTheme } from "@mui/material/styles";
+
+const theme = createTheme({
```
Comment (Satttoshi) [inline pages/_app.tsx:6]: I am new to MUI, I assume here i can override the standard MUI settings

File: pages/index.tsx [hunk 4]
```diff
@@ -1,33 +1,29 @@
...
+import WikiUpdater from "../components/WikiUpdater";
 
 const Home: NextPage = () => {
   return (
-    <>
-      <StyledMain>
-        <Styledh1>Github Wiki Generator</Styledh1>
+    <Grid container p={6}>
+      <Grid item xs={12}>
+        <Typography variant="h4" component="h1" my={2} align="center">
```
Comment (Satttoshi) [inline pages/index.tsx:12]: Dont know why you have to go crazy with Typography here, if there is basically only 1 x h1 on the page :D

File: pages/index.tsx [hunk 5]
```diff
@@ -1,33 +1,29 @@
...
+import Grid from "@mui/material/Grid";
+import MarkDownOutput from "../components/MarkdownOutput";
+import WikiUpdater from "../components/WikiUpdater";
 
 const Home: NextPage = () => {
   return (
-    <>
-      <StyledMain>
-        <Styledh1>Github Wiki Generator</Styledh1>
+    <Grid container p={6}>
```
Comment (Satttoshi) [inline pages/index.tsx:10]: Is this grid thingy something to faster create Grid Boxes with MUI?

File: services/clipboard/index.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export default function copyToClipBoard(text: string) {
+  navigator.clipboard.writeText(text).then(
+    () => {
+      alert(`copied markdown to the clipboard`);
```
Comment (Satttoshi) [inline services/clipboard/index.ts:4]: Annoying popup, I don't like it

File: services/github/index.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    // Commit and push the changes
+    await git.cwd(`./wikis/${repo}`);
+    await git.add("./*");
+    await git.commit("Updated the wiki");
+    await git.push("origin", "master");
+
+    console.log(`Wiki page "${params.wikiPage}" has been updated.`);
+
+    // Clear the repository immediately after pushing
+    await fs.rm(`./wikis/${repo}`, { recursive: true, force: true });
```
Comment (Satttoshi) [inline services/github/index.ts:41]: This is for deleting the local.md file right?
Do fs methods work on vercel?

File: services/github/index.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    await fs.writeFile(
+      path.join(`./wikis/${repo}`, `${params.wikiPage}.md`),
+      params.wikiContent
+    );
+
+    // Commit and push the changes
+    await git.cwd(`./wikis/${repo}`);
+    await git.add("./*");
+    await git.commit("Updated the wiki");
+    await git.push("origin", "master");
```
Comment (Satttoshi) [inline services/github/index.ts:36]: I assume that `"main"` should be here. Does it make a difference thought?

File: services/github/index.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+export default async function updateWiki(params: WikiParams): Promise<void> {
+  try {
+    const git: SimpleGit = simpleGit();
+
+    // Clone the wiki
+    const wikiUrl: string = ssh
+      ? `git@github.com:${owner}/${repo}.wiki.git`
+      : `https://${GH_ACCESS_TOKEN}@github.com/${owner}/${repo}.wiki.git`;
+
+    await git.clone(wikiUrl, `./wikis/${repo}`);
```
Comment (Satttoshi) [inline services/github/index.ts:24]: Something wrong here. I cant upload wiki to my GitHub wiki (I used the github wiki generator as a test wiki). 

Some console output ... and yes i did enter the config data correctly :joy:
image

File: components/WikiUpdater.tsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      <Button
+        variant="outlined"
+        type="submit"
+        disabled={!wikiEntry}
+        startIcon={<UpgradeIcon />}
+        endIcon={isLoading && <CircularProgress size={20} />}
+      >
+        Update Wiki
+      </Button>
+      {data && <Alert severity="success">{data.message}</Alert>}
```
Comment (Satttoshi) [inline components/WikiUpdater.tsx:47]: This returns success, even though the upload failed in the backend. The success currently only checks if data to push exists

File: components/MarkdownOutput.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+          onChange={async (value) => {
+            console.log(value);
+            alterWikiEntry(value);
+          }}
+        />
+      </Stack>
+      <Button
+        startIcon={<ContentCopyIcon />}
+        onClick={() => copyToClipBoard(data.result)}
+      >
```
Comment (Satttoshi) [inline components/MarkdownOutput.tsx:42]: On my 16:9 1920x1080 monitor on my opened chrome, the Button is cutted, no gud UX here. I assume its due to Macbooks usually have more height in relation to width.

image

**Review by Satttoshi — CHANGES_REQUESTED** [review]
Thanks for fueling up your Bulldozer and for flattening the grounds of my files ... within a single commit xD.

I learned a thing or 2 here, I appreciate that alot. :pray:

However, there needs to be some things to be fixed before i can merge your feature. Because i couldn't get the update wiki thing to run, i would like you to check it out.

Wrote some comments, please check them out if you got sum time :v:

File: hooks/useFetch.ts (hunk 1 above)
Comment (doemser) [reply hooks/useFetch.ts:29]: the route slugs also serve as key for fetch state.

File: package.json (hunk 2 above)
Comment (doemser) [reply package.json:14]: isnt it?

File: pages/_app.tsx (hunk 3 above)
Comment (doemser) [reply pages/_app.tsx:6]: Yep exactly

File: pages/index.tsx (hunk 4 above)
Comment (doemser) [reply pages/index.tsx:12]: MUI hast the Typography component and I see no reason in not utilizing it. I need to do something with it, why shouldnt I use it? Also Typography is used for the other headings in the app, at at least 3 places.

File: pages/index.tsx (hunk 5 above)
Comment (doemser) [reply pages/index.tsx:10]: It is the component MUI offers for Grid based Layouts, there is also Stack for flexbox Layouts. the cool thing about it are the easy to set breakpoints:

https://mui.com/material-ui/customization/breakpoints/

Which I also utilize. xs={12} means 100% of the available width on lowest screen size, xs={6} means half of the available space. 12 ist the maximum, but we can chat about it in person if you want to,.

File: services/clipboard/index.ts (hunk 6 above)
Comment (doemser) [reply services/clipboard/index.ts:4]: You can just change it to how you like it, I copied it from another file and forgot to change it, you were utilizing some useRef madness there :D

File: services/github/index.ts (hunk 7 above)
Comment (doemser) [reply services/github/index.ts:41]: In order to write the wiki page the respective repository is cloned, the wiki page is written, pushed and afterwards the clone repository is deleted again to leave no mess

File: services/github/index.ts (hunk 8 above)
Comment (doemser) [reply services/github/index.ts:36]: I dont think it does the same, I did not even recognize it, I will try later to change it.

File: services/github/index.ts (hunk 9 above)
Comment (doemser) [reply services/github/index.ts:24]: hmm using SSH or HTTPS?

File: components/WikiUpdater.tsx (hunk 10 above)
Comment (doemser) [reply components/WikiUpdater.tsx:47]: true, trash alert

File: components/MarkdownOutput.tsx (hunk 11 above)
Comment (doemser) [reply components/MarkdownOutput.tsx:42]: sure you werent scrolling the vscode? At my machine it has more then enough space at the bottom

File: services/github/index.ts (hunk 9 above)
Comment (Satttoshi) [reply services/github/index.ts:24]: SSH

Comment (doemser) [reply services/github/index.ts:24]: think you just did not allow your repository to have a wiki yet, you can do it over the repositories settings.

**Pull request merged**